.pre-commit-hooks.yaml

```diff
@@ -0,0 +1,14 @@
+- id: terraform-docs
+  name: terraform-docs
+  description: Generate documentation for Terraform modules
+  language: golang
+  entry: terraform-docs
+  pass_filenames: false
+  types: [terraform]
```

Comment:
I'm not too familiar with building pre-commit hooks, does this mean it's going to build the binary on the fly for each execution? How does it know how to build it?
Is there an option to use whatever is installed in `PATH` instead?

Reply:
I spoke too soon! [This](https://pre-commit.com/#golang) is how it works, right?

Also it seems there is a [system](https://pre-commit.com/#system) language. Can we define one additional hook for it? Something like this:

```yaml
- id: terraform-docs
  language: system
  entry: terraform-docs

- id: terraform-docs-go
  language: golang
  entry: terraform-docs

- id: terraform-docs-docker
  language: docker
  entry: terraform-docs
```

Reply:
Yes, that is how it works. Admittedly I cannot get it to build on my box, I get:

```
build github.com/terraform-docs/terraform-docs: cannot load embed: malformed module path "embed": missing dot in first path element
```

Is this something you've seen before? I'm not a Go developer so I just used Docker instead 😅

The `system` hook is generally for unsupported languages or languages that don't have virtual env support. What would be the reason to add that for a Go-based project instead of using the native support?

Reply:
First of all that error means you don't have go1.16 installed on your machine. Which actually proves my point a bit!

I'm still not sure about best practices of building a git hook, but I imagine most of terraform-docs users don't necessarily have go installed or not the version we depend on (yourself included), but they most probably have terraform-docs installed (with pre-built binaries, brew, or some other way.) So giving folks an option to choose between binary on `PATH` or Docker image or alternatively building from go source.

Reply:
> First of all that error means you don't have go1.16 installed on your machine.

Weird, I thought I tried with 1.16, but I just reinstalled to confirm and it works now 🤷

> I imagine most of terraform-docs users don't necessarily have go installed or not the version we depend on

Right, though there are a few ways to handle this:

1. `language_version` parameter on the hook to enforce a minimum version
2. Use Docker
3. `system` hook

If I add `language_version` to the hook file, I get:
```
An unexpected error has occurred: AssertionError: For now, pre-commit requires system-installed golang
```
So I guess that's out 😄 

I suppose adding the `system` is probably the next best option, until Golang 1.16+ is more widely used. The downside is users have one more thing to have to install locally, which goes against the philosophy of pre-commit, but I prefer giving people that choice. 👍

Reply:
That's a valid point, and since we're considering a standalone repository for the hook itself, why not a third option!

- have `golang` and `docker` hooks reside in `terraform-docs/terraform-docs` (this repo has the code to build anyways)
- have `system` hook reside in `terraform-docs/pre-commit-hook` (specially that this repo won't be able to build at all)

But that might complicates things a bit more than it should be, but we're giving clear choices for folks out there however they prefer to utilize this hook.

Reply:
I wouldn't spread it across two repos.

As a pre-commit user, I generally know to look for a `.pre-commit-hooks.yaml` in the repo of a linter I'm considering adopting. The presence of such a file is an immediate adoption boost for me, because I only need to add (for example):

```
  - repo: git://github.com/pre-commit/pre-commit-hooks
    rev: v3.2.0
    hooks:
      - id: end-of-file-fixer
```

To my existing hooks, and boom 💥, I'm already using the tool. I don't install anything extra (except maybe Golang if I don't have it, and even then that's why there's a Docker option), don't have to look across multiple repos, it just works™️.

Most repos I've seen only have one hook, they don't offer a Docker or `system` option. I think it's great that this tool has those options, but it decreases discoverability IMO to spread it out.